utils/measure/measure/util/measure_util.py

```diff
@@ -21,21 +26,113 @@ def __init__(self, power_meter: PowerMeter, config: MeasureConfig) -> None:
+                                      support voltage measurements.
+
+        Error handling:
+            - If voltage measurements are not supported, the program will terminate with an appropriate
+              error message, if measuring power consumption and a dummy load resistive value is set.
+            - For resistance measurements, voltage values below 1 are treated as invalid, and the program
+              exits to avoid incorrect calculations.
+            - Ignores single measurements of <= 0 W.
+        """
+        _LOGGER.info("Measuring average %s over %s seconds", "resistance" if measure_resistance else "power", duration)
         start_time = time.time()
         readings: list[float] = []
+
+        first_measurement = True
+
         while (time.time() - start_time) < duration:
+            if first_measurement:
+                first_measurement = False
+            else:
+                # sleep time exceeds duration
+                if not ((time.time() - start_time + self.config.sleep_time) < duration):
+                    break
+                time.sleep(self.config.sleep_time)
+
+            if measure_resistance:
+                result = self.power_meter.get_power(include_voltage=True)
+                power, voltage = result.power, result.voltage
+
+                if voltage < 1:
+                    _LOGGER.error("Error during measurement: Voltage measurement returned zero. Aborting measurement.")
+                    exit(1)
+
+                if round(power, 2) == 0:
+                    _LOGGER.warning("Invalid measurement: power: %.2f W, voltage: %.2f", power, voltage)
+                    continue
+
+                resistance = round((voltage**2) / power, 4)
+                readings.append(resistance)
+                _LOGGER.debug(
+                    "Measured resistance: %.2f Ω; measured power: %.2f W, voltage: %.2f",
+                    resistance,
+                    power,
+                    voltage,
+                )
+                continue
+
+            if self.dummy_load_value:  # measurement with dummy load
+                # Validate power meter is capable of measuring voltage
+                self._validate_voltage_support()
+
+                result = self.power_meter.get_power(include_voltage=True)
+
+                power, voltage = result.power, result.voltage
+
+                if voltage < 1:
+                    _LOGGER.error("Error during measurement: Voltage measurement returned zero. Aborting measurement.")
+                    exit(1)
+
+                dummy_power = (voltage**2) / self.dummy_load_value
+                power -= dummy_power
+
+                if round(power, 2) <= 0:
+                    _LOGGER.warning(
+                        "Invalid measurement after subtracting dummy load consumption. Calculated consumption: %.2f W; ignoring",
+                        power,
+                    )
+                    continue
```

Comment:
This code is duplicated almost completely at line 178 - 188.
see how we can make this dry to keep the code more readable and maintainable.

Reply:
Agreed, both functions could be merged. I initially didn't even catch that there are two measure functions, I thought there's just one. So I added the same functionality to take_measurement() later.

But I don't see how this can realistically be achieved in the scope of this PR.

I could offload the calculation part, however this would introduce a new private method for just two SLOC. I don't think that's worth the effort.

Better would be to leave it be, for the time being and merge them later have a single measure function which has either a count or a time limit.

Reply:
The cleanest solution here would be to have it always loop over number of measurements, and the time based function is called, it's converted via the sleep time set to numbers of measurement and then the other function is called.